shared-module/usb_cdc/Serial.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Dan Halbert for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "lib/utils/interrupt_char.h"
+#include "shared-module/usb_cdc/Serial.h"
+#include "supervisor/shared/tick.h"
+
+#include "tusb.h"
+
+size_t common_hal_usb_cdc_serial_read(usb_cdc_serial_obj_t *self, uint8_t *data, size_t len, int *errcode) {
+    if (self->timeout < 0.0f) {
+        while (tud_cdc_n_available(self->idx) < len) {
+            RUN_BACKGROUND_TASKS;
+            if (mp_hal_is_interrupted()) {
+                return 0;
+            }
+        }
+    } else if (self->timeout > 0.0f) {
+        uint64_t timeout_ms = self->timeout * 1000;
+        uint64_t start_ticks = supervisor_ticks_ms64();
+        while (tud_cdc_n_available(self->idx) < len &&
+               supervisor_ticks_ms64() - start_ticks <= timeout_ms) {
+            RUN_BACKGROUND_TASKS;
+            if (mp_hal_is_interrupted()) {
+                return 0;
+            }
+        }
+    }
+    // Timeout of 0.0f falls through to here with no waiting or unnecessary calculation.
+    return tud_cdc_n_read(self->idx, data, len);
```

Comment:
Shouldn't we read as we can? That will allow `len` here to be greater than the USB buffer sizes.

Reply:
If I'm not waiting at all, then once I read what's in the buffer, the rest won't be available immediately, but will need to be inserted in to the TinyUSB buffer.

Reply:
Fixed to read as much as possible, including multiple buffer loads, in the time allowed.